### PR TITLE
Helm chart: make more fields configurable for the deployment, daemonset and storage class

### PR DESCRIPTION
**Pull request by Misteur-Z**
### Helm chart

**Is this a bug fix or adding new feature?**

Start of my work: pod annotations not being specific to the Controller deployment or the Node daemonset (bug)
==> new feature: make everything properly configurable

**What is this PR about? / Why do we need it?**

We should be able to set annotations, resources, nodeSelector, tolerations, affinity, etc. specifically. Improvement from https://github.com/kubernetes-sigs/aws-efs-csi-driver/pull/398 not fully addressing these issues.

I tried to find an elegant solution to this issue but it's very hard not to create breaking changes in the Helm values file.

**What testing is done?** 

Tested in AWS EKS 1.19.6 with Helm 3.5.3

**What else?**

I added a changelog for the Helm chart, could be useful.

### CODE REVIEW

File: charts/aws-efs-csi-driver/templates/controller-deployment.yaml
Comment (Misteur-Z) [inline charts/aws-efs-csi-driver/templates/controller-deployment.yaml:23]: `.Values.podAnnotations` was actually unused => not working (found out while using `kube2iam`)

**Comment by Misteur-Z** [comment]
I signed it

**Comment by wongma7** [comment]
Heads up, you might have to do some rebasing, I might merge a bunch of 1.2.x changes first in the next few days, however in principle I approve of this PR.

**Comment by Misteur-Z** [comment]
I signed it

**Comment by Misteur-Z** [comment]
* Seems good to go, let me test the Chart again just in case
* I didn't update the Helm chart changelog all the way to version `1.0.0`

**Comment by Misteur-Z** [comment]
I signed it

**Comment by Misteur-Z** [comment]
* Seems good to go, tested the Chart again in EKS
* I didn't update the Helm chart changelog all the way to version `1.0.0`

File: charts/aws-efs-csi-driver/Chart.yaml
Comment (wongma7) [inline charts/aws-efs-csi-driver/Chart.yaml:5]: do we need to bump the major version? I'm not really sure about best practice but I think yes?

File: charts/aws-efs-csi-driver/Chart.yaml
Comment (wongma7) [inline charts/aws-efs-csi-driver/Chart.yaml:6]: what change is requiring 1.18?

Comment (wongma7) [reply charts/aws-efs-csi-driver/Chart.yaml:6]: oh never mind, I see it's because of v1 csidriver in the changelog.

This complicates matters because technically the driver works with 1.17+ but now the primary install method requires 1.18+.

looks like ebs did the same change though so...https://github.com/kubernetes-sigs/aws-ebs-csi-driver/pull/836

File: charts/aws-efs-csi-driver/Chart.yaml
Comment (wongma7) [reply charts/aws-efs-csi-driver/Chart.yaml:5]: In the apst, we have bumped minor version to signal that a new efs driver version is out.

So e.g. currently it so happens that 1.2.x correspondx to efs driver 1.2.y.

That's why I think we need a major version bump

Comment (Misteur-Z) [reply charts/aws-efs-csi-driver/Chart.yaml:5]: I am creating breaking changes in the values file, that's why I thought about bumping the major version. I would say most helm charts versions don't match the app version.

File: charts/aws-efs-csi-driver/templates/_helpers.tpl
Comment (Misteur-Z) [inline charts/aws-efs-csi-driver/templates/_helpers.tpl:54]: `serviceAccount` is controller specific

**Comment by Misteur-Z** [comment]
Merged master again:
- fixed `logLevel` not being specific to either the controller deployment or node daemonset
- moved `serviceAccount` into `controller` since it's specific to its deployment

**Comment by wongma7** [comment]
thanks, everything lgtm but I'm just hesitant about the versioning, overall I'm satisfied that this PR is making the chart more consistent and clearly partitioning values along controller/node lines.

AFter it merges, I won't make the release immediately, I need to test especially the service account and RBAC wiring, since we've had longstanding issues there: https://github.com/kubernetes-sigs/aws-efs-csi-driver/issues/347 https://github.com/kubernetes-sigs/aws-efs-csi-driver/issues/421

File: charts/aws-efs-csi-driver/Chart.yaml
Comment (Misteur-Z) [reply charts/aws-efs-csi-driver/Chart.yaml:5]: Good idea since there is a lot of potential breaking changes! Let's bump to `2.0.0` to make it significant.

**Comment by wongma7** [comment]
@kbasv @ayberk  could you guys help me audit the helm chart and suggest any changes? This PR is making some breaking changes and bumping to 2.0.0 so as @Misteur-Z said, if there's anything else we want to change, now is the time to do it!

**Comment by wongma7** [comment]
Also @krmichel if you wouldn't mind reviewing, yo uare a helm chart maintainer so would appreciate your eyes on this!!

File: charts/aws-efs-csi-driver/CHANGELOG.md
Comment (ayberk) [inline charts/aws-efs-csi-driver/CHANGELOG.md:10]: Can we have a specific section for "Breaking Changes"? Ideally we don't want the users to go read the template and figure out every single change they need to make.

File: charts/aws-efs-csi-driver/templates/controller-deployment.yaml
Comment (ayberk) [inline charts/aws-efs-csi-driver/templates/controller-deployment.yaml:50]: Can we move `pullPolicy` to their own fields under `node` and `controller`? Just to be consistent.

**Review by ayberk — CHANGES_REQUESTED** [review]

File: charts/aws-efs-csi-driver/templates/controller-deployment.yaml
Comment (wongma7) [reply charts/aws-efs-csi-driver/templates/controller-deployment.yaml:50]: that would be less consistent cuz it would break the pattern of image structs all having the same fields pullPolicy, tag , & repository. The top-level `image` is the driver image which should be the same for both node and controller.

File: charts/aws-efs-csi-driver/values.yaml
Comment (wongma7) [inline charts/aws-efs-csi-driver/values.yaml:79]: while we r here, default should be lower, like 2: https://github.com/kubernetes-sigs/aws-ebs-csi-driver/issues/854

File: charts/aws-efs-csi-driver/Chart.yaml
Comment (krmichel) [reply charts/aws-efs-csi-driver/Chart.yaml:5]: They did but I don't think it was handled correctly there.  They didn't change the kube version dependency in the chart.  I don't think it requires that the chart only work on 1.18+

File: charts/aws-efs-csi-driver/templates/csidriver.yaml
Comment (krmichel) [inline charts/aws-efs-csi-driver/templates/csidriver.yaml:1]: This would be better handled like this and then you wouldn't need to increase the minimum kube version.
https://github.com/kubernetes-sigs/aws-ebs-csi-driver/pull/856/files#diff-e167822724b087c64ed35c6f75c8af5d054a981e00b59488145796c69c6c611fR1

`apiVersion: {{ ternary "storage.k8s.io/v1" "storage.k8s.io/v1beta1" (semverCompare ">=1.18.0-0" .Capabilities.KubeVersion.Version) }}`

**Review by krmichel — COMMENTED** [review]
> Also @krmichel if you wouldn't mind reviewing, yo uare a helm chart maintainer so would appreciate your eyes on this!!

@wongma7 I will take a look at this.  I have started a review but don't have time to finish it right now.  Please don't merge until I have had a chance to finish my review.  There are a number of breaking changes that could be avoided if the changes were made in two steps.  Is that a concern that I should make comments about during my review or should I just assume that breaking changes are fine?

File: charts/aws-efs-csi-driver/CHANGELOG.md
Comment (Misteur-Z) [reply charts/aws-efs-csi-driver/CHANGELOG.md:59]: Tried to explain breaking changes in b1c989de876b5cc719153f51dbe3a831a0a731ed

File: charts/aws-efs-csi-driver/templates/csidriver.yaml
Comment (Misteur-Z) [reply charts/aws-efs-csi-driver/templates/csidriver.yaml]: Fixed in b1c989de876b5cc719153f51dbe3a831a0a731ed

File: charts/aws-efs-csi-driver/values.yaml
Comment (Misteur-Z) [reply charts/aws-efs-csi-driver/values.yaml]: Lowered to 2 in b1c989de876b5cc719153f51dbe3a831a0a731ed

**Comment by Misteur-Z** [comment]
Merged master, addressed all the comments.

File: charts/aws-efs-csi-driver/values.yaml
Comment (krmichel) [inline charts/aws-efs-csi-driver/values.yaml:99]: move `{}` to this line to make the example easier to read

File: charts/aws-efs-csi-driver/values.yaml
Comment (krmichel) [inline charts/aws-efs-csi-driver/values.yaml:52]: move `{}` to this line to make the example easier to read

File: charts/aws-efs-csi-driver/values.yaml
Comment (krmichel) [inline charts/aws-efs-csi-driver/values.yaml:80]: move `{}` to this line to make the example easier to read

File: charts/aws-efs-csi-driver/values.yaml
Comment (krmichel) [inline charts/aws-efs-csi-driver/values.yaml:91]: move `{}` to this line to make the example easier to read

File: charts/aws-efs-csi-driver/CHANGELOG.md
Comment (krmichel) [inline charts/aws-efs-csi-driver/CHANGELOG.md:5]: You don't mention moving nodeSelector and resources under node.  It is probably also worth noting that you fixed a bug where the controller was using podAnnotations and tolerations from the node section and is now using the correct values from the controller section.  It might be worth putting a section in the readme under the helm instructions that point here for upgrading from versions prior to 2.0.0

File: charts/aws-efs-csi-driver/templates/controller-deployment.yaml
Comment (krmichel) [inline charts/aws-efs-csi-driver/templates/controller-deployment.yaml:39]: It might be worth considering adding the ability to set topologySpreadConstraints while this change is being made.  This is untested but might look something like this.

```
      {{- with .Values.controller.topologySpreadConstraints }}
      {{- $tscLabelSelector := dict "labelSelector" (dict "matchLabels" (dict "app" "efs-csi-controller")) }}
      topologySpreadConstraints:
        {{- range . }}
        - {{ mergeOverwrite . $tscLabelSelector | toJson }}
        {{- end }}
      {{- end }}
```

**Review by krmichel — CHANGES_REQUESTED** [review]
Okay, I finished a pass through the whole thing.

File: charts/aws-efs-csi-driver/Chart.yaml
Comment (krmichel) [inline charts/aws-efs-csi-driver/Chart.yaml:1]: While making a breaking change might as well make this version helm API v2.

File: charts/aws-efs-csi-driver/templates/controller-deployment.yaml
Comment (Misteur-Z) [reply charts/aws-efs-csi-driver/templates/controller-deployment.yaml:41]: I wanted to fix this but I didn't want to introduce some breaking change or new issue. Can we let users set `tolerations`?

File: charts/aws-efs-csi-driver/templates/controller-deployment.yaml
Comment (Misteur-Z) [reply charts/aws-efs-csi-driver/templates/controller-deployment.yaml:39]: I am not sure of the implications of changing anything here, are we supporting only Kubernetes running on Linux hosts?

File: charts/aws-efs-csi-driver/CHANGELOG.md
Comment (Misteur-Z) [reply charts/aws-efs-csi-driver/CHANGELOG.md:5]: Tried my best to clarify everything, tell me if it's not enough

File: charts/aws-efs-csi-driver/templates/node-daemonset.yaml
Comment (Misteur-Z) [reply charts/aws-efs-csi-driver/templates/node-daemonset.yaml:53]: Test the helm chart this morning, feeding `dnsPolicy` with an empty value will use default `ClusterFirst`

File: deploy/kubernetes/base/controller-serviceaccount.yaml
Comment (Misteur-Z) [reply deploy/kubernetes/base/controller-serviceaccount.yaml]: Not related to the helm chart

File: charts/aws-efs-csi-driver/templates/controller-deployment.yaml
Comment (Misteur-Z) [reply charts/aws-efs-csi-driver/templates/controller-deployment.yaml:41]: Switching to let the user define the whole `tolerations` array in values file.

File: charts/aws-efs-csi-driver/templates/controller-deployment.yaml
Comment (Misteur-Z) [reply charts/aws-efs-csi-driver/templates/controller-deployment.yaml:39]: Not adding `topologySpreadConstraints` for now, sticking with the current `nodeSelector` and `priorityClassName`.

**Comment by Misteur-Z** [comment]
Feels good to go @krmichel, not sure if there is anything to fix in `deploy/` or elsewhere @wongma7

File: charts/aws-efs-csi-driver/templates/node-daemonset.yaml
Comment (krmichel) [inline charts/aws-efs-csi-driver/templates/node-daemonset.yaml:62]: The daemonset and the deployment should behave differently.  Removing the operator: Exists from the daemonset I think would be surprising to users.  I think a daemonset needs some sort of default tolerations since it is supposed to be running on "every" node.  I think a reasonable solution might be to have the operator: Exists as a default in the values.yaml and then if a user wants they have a method of getting rid of that toleration.

```
node:
  ...
  tolerations:
    - operator: Exists
```

**Comment by krmichel** [comment]
> Feels good to go @krmichel, not sure if there is anything to fix in `deploy/` or elsewhere @wongma7

I think it looks pretty good except dealing with the daemonset tolerations.  Once that is addressed I should be good with it.

File: charts/aws-efs-csi-driver/templates/node-daemonset.yaml
Comment (Misteur-Z) [reply charts/aws-efs-csi-driver/templates/node-daemonset.yaml:62]: Fixed: https://github.com/kubernetes-sigs/aws-efs-csi-driver/pull/406/commits/a2aa0794e2ddd4921e068fc044d201a67e86b662

**Review by krmichel — APPROVED** [review]
looks good

**Comment by wongma7** [comment]
/hold cancel

Sorry Ive been quite busy for last 2 weeks but will do final pass and merge soon

**Comment by wongma7** [comment]
/hold cancel

Sorry Ive been quite busy for last 2 weeks but will do final pass and merge soon

File: charts/aws-efs-csi-driver/templates/controller-deployment.yaml
Comment (wongma7) [reply charts/aws-efs-csi-driver/templates/controller-deployment.yaml:39]: works for me, I don't think topologySpreadConstraints would make sense on node, we can always add later to controller if somebody asks for it.

**Comment by wongma7** [comment]
/approve

thank you very much. Appreciate the patience, reviewer bandwidth is quite low... do you mind squashing the commits into one or two? using git rebase -i or something similar. It will make it easier to backport to release-1.2 branch  ( which to be fair I am not 100% sure we should do yet). Merging this PR WON'T immediately release the 2.0.0 chart, somebody still needs to do some manual stuff to release the chart.

**Comment by Misteur-Z** [comment]
Squashed into a single commit

**Comment by rmgpinto** [comment]
@wongma7 can you release the 2.0.0 chart?
Thanks

**Comment by wongma7** [comment]
@rmgpinto it's coming https://github.com/kubernetes-sigs/aws-efs-csi-driver/pull/458

**Pull request merged**
